code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/process/ServiceActivate.java

```diff
@@ -0,0 +1,246 @@
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+
+@Named
+public class ServiceActivate extends AbstractObjectProcessHandler {
+    @Inject
+    JsonMapper jsonMapper;
+
+    @Inject
+    NetworkDao ntwkDao;
+
+    @Inject
+    StorageService storageService;
+
+    @Inject
+    ObjectManager objectManager;
+
+    @Inject
+    GenericMapDao mapDao;
+    
+    @Inject
+    ResourceMonitor resourceMonitor;
+
+    @Inject
+    ServiceDiscoveryServer serviceDiscoveryServer;
+
+    @Inject
+    GenericResourceDao resourceDao;
+
+    @Inject
+    ProcessProgress progress;
+
+    @Inject
+    ServiceDiscoveryServer sdServer;
+
+    @Override
+    public String[] getProcessNames() {
+        return new String[] { ServiceDiscoveryConstants.PROCESS_SERVICE_ACTIVATE };
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public HandlerResult handle(ProcessState state, ProcessInstance process) {
+        Service service = (Service) state.getResource();
+        List<Integer> consumedServiceIds = new ArrayList<>();
+        boolean activateConsumedServices = DataAccessor.fromMap(state.getData()).withScope(ServiceActivate.class)
+                .withKey(ServiceDiscoveryConstants.FIELD_ACTIVATE_CONSUMED_SERVICES).withDefault(false)
+                .as(Boolean.class);
+        if (activateConsumedServices) {
+            // the order is determined only based on volumes_from (service) attribute
+            // we don't look at service links because the services get linked via DNS updates,
+            // and DNS update logic is going to be handled separately from service.activate
+            List<? extends Integer> services = (List<? extends Integer>) DataUtils.getFields(service).get(
+                    ServiceDiscoveryConfigItem.VOLUMESFROMSERVICE.getRancherName());
+            if (services != null) {
+                consumedServiceIds.addAll(services);
+            }
+        }
+
+        int scale = DataAccessor.field(service, ServiceDiscoveryConstants.FIELD_SCALE,
+                jsonMapper,
+                Integer.class);
+        progress.init(state, sdServer.getWeights(scale + consumedServiceIds.size(), 100));
+
+        if (activateConsumedServices) {
+            activateConsumedServices(consumedServiceIds, state.getData());
+        }
+
+        createServiceInstances(service, scale);
+        return null;
+    }
+
+    private void activateConsumedServices(List<? extends Integer> consumedServiceIds,
+            Map<String, Object> data) {
+        List<Service> consumedServices = new ArrayList<>();
+        if (consumedServiceIds != null) {
+            for (Integer consumedServiceId : consumedServiceIds) {
+                Service consumedService = objectManager.findOne(Service.class, SERVICE.ID, consumedServiceId,
```

Comment:
Minor: Can we add a comment line to note that we should probably bulk load the services in the future.  Definitely don't need to do it now but I think it's worth the optimization in the future.